include/drivers/can.h

```diff
@@ -509,28 +562,196 @@ static inline void z_impl_can_detach(const struct device *dev, int filter_id)
+		       uint32_t bitrate);
+
+/**
+ * @brief Set the controller to the given mode
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param mode   Operation mode
  *
  * @retval 0 If successful.
  * @retval -EIO General input / output error, failed to configure device.
  */
-__syscall int can_configure(const struct device *dev, enum can_mode mode,
-			    uint32_t bitrate);
+__syscall int can_set_mode(const struct device *dev, enum can_mode mode);
 
-static inline int z_impl_can_configure(const struct device *dev,
-				       enum can_mode mode,
-				       uint32_t bitrate)
+static inline int z_impl_can_set_mode(const struct device *dev,
+				      enum can_mode mode)
 {
 	const struct can_driver_api *api =
 		(const struct can_driver_api *)dev->api;
 
-	return api->configure(dev, mode, bitrate);
+	return api->set_mode(dev, mode);
 }
 
+/**
+ * @brief Configure timing of a host controller.
+ *
+ * The second parameter timing_data is only relevant for CAN-FD.
+ * If the controller does not support CAN-FD or the FD mode is not enabled,
+ * this parameter is ignored.
+ *
+ * @param dev         Pointer to the device structure for the driver instance.
+ * @param timing      Bus timings
+ * @param timing_data Bus timings for data phase (CAN-FD only)
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error, failed to configure device.
+ */
+__syscall int can_set_timing(const struct device *dev,
+			     const struct can_timing *timing,
+			     const struct can_timing *timing_data);
+
+static inline int z_impl_can_set_timing(const struct device *dev,
+					const struct can_timing *timing,
+					const struct can_timing *timing_data)
+{
+	const struct can_driver_api *api =
+		(const struct can_driver_api *)dev->api;
+
+	return api->set_timing(dev, timing, timing_data);
+}
+
+/**
+ * @brief Set the bitrate of the CAN-FD controller
+ *
+ * The sample point is set to the CiA DS 301 reccommended value of 87.5%
+ *
+ * @param dev          Pointer to the device structure for the driver instance.
+ * @param bitrate      Desired arbitration phase bitrate
+ * @param bitrate_data Desired data phase bitrate
+ *
+ * @retval 0 If successful.
+ * @retval -EINVAL bitrate cannot be reached.
+ * @retval -EIO General input / output error, failed to set bitrate.
+ */
+static inline int can_set_bitrate(const struct device *dev,
+				  uint32_t bitrate,
+				  uint32_t bitrate_data)
+{
+	struct can_timing timing;
+#ifdef CONFIG_CAN_FD_MODE
+	struct can_timing timing_data;
+#endif
+	int ret;
+
+	ret = can_calc_timing(dev, &timing, bitrate, 875);
+	if (ret < 0) {
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_CAN_FD_MODE
+	ret = can_calc_timing_data(dev, &timing_data, bitrate_data, 875);
+	if (ret < 0) {
+		return -EINVAL;
+	}
+
+	return can_set_timing(dev, &timing, &timing_data);
+#else
+	return can_set_timing(dev, &timing, NULL);
+#endif /* CONFIG_CAN_FD_MODE */
```

Comment:
This forgets to set timing.sjw and timing_data.sjw before calling can_set_timing()

Reply:
@alexanderwachter If you have time, could you have a look pls? ^^
Should I open an issue?

Reply:
@talih0  It's a bug, yes. Can you open an issue?